providers/hns/hns_roce_u.c

```diff
@@ -85,6 +87,17 @@ static const struct verbs_context_ops hns_common_ops = {
 	.destroy_ah = hns_roce_u_destroy_ah,
 };
 
+static bool get_param_config(const char *var)
+{
+	char *env;
+
+	env = getenv(var);
```

Comment:
Didn't we talk in ML that environmental variable is not the best way to configure library?

Reply:
Hi leon,
Yes, you have point that.But could you please tell me

Reply:
Hi leon,
Yes, you have pointed that, but sorry for failing to find another way to configure library. Could you please share more information? Thank you :)

Reply:
In order to answer this question properly, I need to understand why and what do you mean by setting loopback to enable/disable.

First, we have IBV_QP_CREATE_BLOCK_SELF_MCAST_LB, second we have mlx5dv_create_qp() for more granular solution, third we have loopback in librdmacm,
https://github.com/linux-rdma/rdma-core/blob/master/Documentation/librdmacm.md

Reply:
Hi, leon Romanovsky
   This is the only version we currently think of that supports loopback.  In fact, I have always had a question for loopback.

if the rdma device want to supports loopback capability, it should be set in QPC(RC and other transport) or WQE(UD) for user.
I think that the indicator bit should expose to the user.  Also, I am learning the IB protocol chapter 11.
For example:
the chapter 11.2.2.1 CREATE ADDRESS HANDLE
 ...
 ...
 ...

• Address vector, for UD transports only, containing:
• Service level.
• Send Global Routing Header Flag.
• Destination LID. If destination is in same subnet, LID = final
destination; otherwise LID = router LID. DLID = 0xFFFF is illegal
except when the destination QP is QP0.
• Loopback Indicator (if the HCA supports Loopback Indicator
on Address Handles). DLID and Loopback Indicator are mutually
exclusive.

I think that the ULP should provider a paramter loopback for user.  Is this understood?
However, I didn't find it in anywhere.

Reply:
Because hipxx hardware set the loopback indicator bit in QPC and UD WQE. When the below condition established, the hardware will generate loopback packet in mac level.
1. smac == dmac
2. dgid == sgid
3. or the QPC.loopback indicator bit will be set for RC or UDWQE.loopback indicator bit will be set for UD

Reply:
First, you need to provide loopback info through query HCA.

> o10-6.2-1.1: If the CI supports the Loopback Indicator, the CI shall report it
> through the Query HCA Verb. The CI shall accept the Loopback Indicator
> as an input modifier to the Create Address Handle, Modify Address 
> Handle, Modify QP and Modify EEC verbs. The CI shall support the Loopback
> Indicator as an output modifier of the Query QP, Query EEC and Poll
> for Completion verbs.

IMHO, providers don't support this indicator, because the value is little.

Reply:
Hi, leon
   I am learned the IB protocol.  In protocol 11.2.1.2 QUERY HCA, Clearly mentioned support for query loopback. However, we can't find the output definition for ib_device_attr in ib_query_device verbs.  Besides, the wc have no the loopback member  when poll cq.
As a result, how to tell user the hw generate a loopback packet?

  I am also learned the  usage of IBV_QP_CREATE_BLOCK_SELF_MCAST_LB. I think that it is indicator the hw support multicast loopback. But, the hw is only support unicast and need to
support loopback,  should it be?
  At the same time I found, the FL of mlx wqe will be set when smac is equal to dmac.  I think it  is in violation of the IB agreement. Because mac is equal and loopback is the two equal conditions to determine whether to loop back？
Maybe I understand it wrong. I don't understand any other setup methods at the moment.

Reply:
Why existing solutions don't work for you?

Reply:
From the current analysis, I still can't understand the tree methods you suggested can solve.
As use IBV_QP_CREATE_BLOCK_SELF_MCAST_LB, it should be support multicast. however, some hardwares only support unicast.
if use the third way, I think it is global. Is this method consistent with setting input parameters?

Reply:
Hi Leon,
Thank you for your help. We have removed the patch about loopback from this patchset, and we will keep working on it. Do you have more suggestions on other patches?

Reply:
Please close this PR and start a new one, it is hard to see the changes now.
The kernel-headers files should be updated with kernel-header/update script.